src/approaches/operator_learning_approach.py

```diff
@@ -0,0 +1,210 @@
+                part_add_effects = add_effects[i]
+                part_delete_effects = delete_effects[i]
+                if OperatorLearningApproach._unify(
+                        trans_add_effects, trans_delete_effects,
+                        part_add_effects, part_delete_effects)[0]:
+                    # Add to this partition
+                    partition_index = i
+                    break
+            # Otherwise, create a new group
+            if partition_index is None:
+                new_partition = [transition]
+                partitions.append(new_partition)
+                # Get new lifted effects
+                objects = {o for atom in trans_add_effects |
+                           trans_delete_effects for o in atom.objects}
+                objects_lst = sorted(objects)
+                variables = [Variable(f"?x{i}", o.type)
+                             for i, o in enumerate(objects_lst)]
+                sub = dict(zip(objects_lst, variables))
+                add_effects.append({atom.lift(sub) for atom
+                                    in trans_add_effects})
+                delete_effects.append({atom.lift(sub) for atom
+                                       in trans_delete_effects})
+                num_partitions += 1
+            # Add to existing group
+            else:
+                partitions[partition_index].append(transition)
+
+        assert len(add_effects) == len(delete_effects) == len(partitions) == \
+            num_partitions
+        return add_effects, delete_effects, partitions
+
+    @staticmethod
+    def  _learn_preconditions(
+            add_effects: Set[LiftedAtom], delete_effects: Set[LiftedAtom],
+            transitions: List[Transition]) -> Tuple[
+                Sequence[Variable], Set[LiftedAtom]]:
+        for i, (atoms, _, trans_add_effects,
+                trans_delete_effects) in enumerate(transitions):
+            suc, sub = OperatorLearningApproach._unify(
+                trans_add_effects, trans_delete_effects,
+                add_effects, delete_effects)
+            assert suc  # else this transition won't be in this partition
+            # Remove atoms from the state which contain objects not mentioned
+            # in the effects. This cannot handle actions at a distance.
+            objects = {o for atom in trans_add_effects |
+                       trans_delete_effects for o in atom.objects}
+            atoms = {atom for atom in atoms if
+                     all(o in objects for o in atom.objects)}
+            lifted_atoms = {atom.lift(sub) for atom in atoms}
+            if i == 0:
+                variables = sorted(set(sub.values()))
+            else:
+                assert variables == sorted(set(sub.values()))
+            if i == 0:
+                preconditions = lifted_atoms
+            else:
+                preconditions &= lifted_atoms
+
+        return variables, preconditions
+
+    @staticmethod
+    def _unify(
```

Comment:
can we lru_cache this? it's going to be called twice I think -- first during effect clustering, and then again during precondition learning